### PR TITLE
audiobookshelf: init module

### PR DESCRIPTION
**Pull request by wietsedv**
## Description of changes

Add audiobookshelf module. Two previous PRs seem to be outdated/abandoned: #201552 and #245184.

This PR uses the test written by @dit7ya (#245184), but the main module is mostly written by myself.

@rembo10 and @dit7ya, let me know if you want to be added as maintainers.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -18582,6 +18582,12 @@
     githubId = 7121530;
     name = "Wolf Honoré";
   };
+  wietsedv = {
+    email = "wietsedv@proton.me";
+    github = "wietsedv";
+    githubId = 13139101;
+    name = "Wietse de Vries";
+  };
```
Comment (ambroisie) [inline maintainers/maintainer-list.nix:18590]: Please split this into a different commit (`maintainers: add wietsedv`).

File: nixos/modules/misc/ids.nix [hunk 2]
```diff
@@ -356,6 +356,7 @@ in
       rstudio-server = 324;
       localtimed = 325;
       automatic-timezoned = 326;
+      audiobookshelf = 327;
```
Comment (ambroisie) [inline nixos/modules/misc/ids.nix:359]: No need for static uid, those are legacy.

Just create `users.users.audiobookshelf` without it.

File: nixos/modules/misc/ids.nix [hunk 3]
```diff
@@ -666,6 +667,7 @@ in
...
```
Comment (ambroisie) [inline nixos/modules/misc/ids.nix:670]: No need for static uid, those are legacy.

Just declare `users.groups.audiobookshelf` without it.

File: nixos/modules/services/web-apps/audiobookshelf.nix [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    services.audiobookshelf = {
+      enable = mkEnableOption "Audiobookshelf, self-hosted audiobook and podcast server.";
+
+      package = mkOption {
+        description = "Audiobookshelf package to use";
+        default = pkgs.audiobookshelf;
+        defaultText = literalExpression "pkgs.audiobookshelf";
+        example = literalExpression "pkgs.audiobookshelf";
+        type = types.package;
+      };
```
Comment (ambroisie) [inline nixos/modules/services/web-apps/audiobookshelf.nix:19]:
```suggestion
      package = mkPackageOptionMd pkgs "audiobookshelf" { };
```

File: nixos/modules/services/web-apps/audiobookshelf.nix [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+      dataDir = mkOption {
+        description = "Path to Audiobookshelf config and metadata.";
+        default = "/var/lib/audiobookshelf";
+        type = types.path;
+      };
```
Comment (ambroisie) [inline nixos/modules/services/web-apps/audiobookshelf.nix:25]: Is there a reason to have it outside of `/var/lib`?

**Review by ambroisie — CHANGES_REQUESTED** [review]

File: nixos/modules/services/web-apps/audiobookshelf.nix (hunk 5 above)
Comment (wietsedv) [reply nixos/modules/services/web-apps/audiobookshelf.nix]: There is not. I have changed `dataDir` to be a name inside of `/var/lib`. This option is one of those nixpkgs conventions that lack standardization, so I find it difficult to know what to do with it.

**Comment by wietsedv** [comment]
@ambroisie Thanks for the review! I have made your suggested changes.

File: maintainers/maintainer-list.nix
Comment (ambroisie) [inline maintainers/maintainer-list.nix:1]: This should be commited _before_ it is used :-)

**Review by ambroisie — CHANGES_REQUESTED** [review]

File: maintainers/maintainer-list.nix
Comment (wietsedv) [reply maintainers/maintainer-list.nix:1]: Makes sense, I have swapped the commits.

**Review by ambroisie — APPROVED** [review]

File: nixos/modules/services/web-apps/audiobookshelf.nix [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        ExecStart = "${cfg.package}/bin/audiobookshelf --host ${cfg.host} --port ${toString cfg.port}";
+        Restart = "on-failure";
+      };
+    };
+
+    users.users = mkIf (cfg.user == "audiobookshelf") {
+      audiobookshelf = {
+        isSystemUser = true;
+        group = cfg.group;
+        home = "/var/lib/${cfg.dataDir}";
```
Comment (turion) [inline nixos/modules/services/web-apps/audiobookshelf.nix:76]: Maybe you want to add `createHome = true;`? I'm not sure it's necessary, though.

File: nixos/modules/services/web-apps/audiobookshelf.nix [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        description = "User account under which Audiobookshelf runs.";
+        default = "audiobookshelf";
+        type = types.str;
+      };
+
+      group = mkOption {
+        description = "Group under which Audiobookshelf runs.";
+        default = "audiobookshelf";
+        type = types.str;
+      };
```
Comment (turion) [inline nixos/modules/services/web-apps/audiobookshelf.nix:44]: Are you sure this flexibility is needed? `audiobookshelf` is a pretty unique name, and you can reduce complexity if you don't have these options.

File: nixos/modules/services/web-apps/audiobookshelf.nix [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    users.groups = mkIf (cfg.group == "audiobookshelf") {
+      audiobookshelf = { };
+    };
+
+    networking.firewall = mkIf cfg.openFirewall {
+      allowedTCPPorts = [ cfg.port ];
+    };
+  };
+
+  meta.maintainers = with maintainers; [ wietsedv ];
```
Comment (turion) [inline nixos/modules/services/web-apps/audiobookshelf.nix:89]: Feel free to add me as a maintainer :)

File: nixos/tests/audiobookshelf.nix [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+      services.audiobookshelf = {
+        enable = true;
+        port = 1234;
+      };
+    };
+
+  testScript = ''
+    machine.wait_for_unit("audiobookshelf.service")
+    machine.wait_for_open_port(1234)
+    machine.succeed("curl --fail http://localhost:1234/")
```
Comment (turion) [inline nixos/tests/audiobookshelf.nix:21]: A follow-up might be to extend this test such that it creates & downloads some podcasts.

File: nixos/modules/services/web-apps/audiobookshelf.nix [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        description = "Group under which Audiobookshelf runs.";
+        default = "audiobookshelf";
+        type = types.str;
+      };
+
+      openFirewall = mkOption {
+        description = "Open ports in the firewall for the Audiobookshelf web interface.";
+        default = false;
+        type = types.bool;
+      };
```
Comment (turion) [inline nixos/modules/services/web-apps/audiobookshelf.nix:50]: An alternative might be to configure an nginx reverse proxy. But it's ok not do that in this PR here.

File: nixos/modules/services/web-apps/audiobookshelf.nix (hunk 7 above)
Comment (turion) [reply nixos/modules/services/web-apps/audiobookshelf.nix:44]: Ah I saw https://github.com/NixOS/nixpkgs/pull/245184#issuecomment-1740589609, seems like you have a use case for it.

File: nixos/modules/services/web-apps/audiobookshelf.nix (hunk 6 above)
Comment (ambroisie) [reply nixos/modules/services/web-apps/audiobookshelf.nix:76]: `StateDirectory` takes care of creating the directory if needed.

File: nixos/modules/services/web-apps/audiobookshelf.nix (hunk 10 above)
Comment (ambroisie) [reply nixos/modules/services/web-apps/audiobookshelf.nix:50]: IMO since the reverse-proxy is so easy to set-up (one-line), there shouldn't be an option for it in the module.

File: nixos/modules/services/web-apps/audiobookshelf.nix (hunk 7 above)
Comment (ambroisie) [reply nixos/modules/services/web-apps/audiobookshelf.nix:44]: I wonder if we could use `DynamicUser` _with_ `User`/`Group` for this use-case :thinking:.

This is just me wondering aloud, not a change request on this PR.

File: nixos/modules/services/web-apps/audiobookshelf.nix (hunk 10 above)
Comment (turion) [reply nixos/modules/services/web-apps/audiobookshelf.nix:50]: It would still be helpful for beginners. I wouldn't know this line by heart, and I'm not even a beginner anymore (I hope). Some [reverse proxy settings on github](https://github.com/advplyr/audiobookshelf#nginx-reverse-proxy) look sort of straightforward, but I'm still unsure whether you could fit them on a line readably.

Also a reverse proxy setup might include further options like a url.

Comment (ambroisie) [reply nixos/modules/services/web-apps/audiobookshelf.nix:50]: Beginners can read the documentation :-), `services.nginx.virtualHosts.<name>.locations."/".proxyPass = "http://127.0.0.1:<port>";` appears in multiple places.

We have multiple examples of setting-up Nginx for simple port forwarding.

There's no complex configuration to be done (websockets, sub-paths, PHP, auth, etc...), I don't think adding an option is necessary.

Comment (turion) [reply nixos/modules/services/web-apps/audiobookshelf.nix:50]: See e.g. https://search.nixos.org/options?channel=23.05&show=services.xandikos.nginx.enable&from=0&size=50&sort=relevance&type=packages&query=nginx

Comment (turion) [reply nixos/modules/services/web-apps/audiobookshelf.nix:50]: I see your point, but still I disagree. Anyways, we both agree that the present PR here is great without that option.

File: nixos/modules/services/web-apps/audiobookshelf.nix (hunk 8 above)
Comment (turion) [reply nixos/modules/services/web-apps/audiobookshelf.nix:89]: In fact, no need for you to do this. I'll do it the next time I come around touching this service.

**Review by turion — APPROVED** [review]

**Comment by turion** [comment]
@ofborg test audiobookshelf
